### PR TITLE
[bbs] update REPO_WRITE permission check EXP-521

### PR DESCRIPTION
**Pull request by mustard-mh**
## Description

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at fe93d62</samp>

Simplified and improved Bitbucket Server token validation. Removed unused `getPermission` method from `BitbucketServerApi` class.

</details>

## Related Issue(s)

Fixes EXP-521

## How to test

- Join org with BBS ingergration added (TODO)
- Authenticate your BBS account
- Try create workspace with https://bitbucket.gitpod-dev.com/projects/MYP/repos/my-repo/browse it should failed since this repo only admin has permission
- Use admin account to grant admin role to your account
- Try create workspace again, it should work without permission notification
- Revoke your permission, and use admin account to grant repo read permission to your account
- Try create worksapce again, it should work with permission notification, and failed to push

**Unit Tests**
```
# terminal 1
cd components/server
yarn watch

# terminal 2
cd components/server

# get access tokens from 1Password
export GITPOD_TEST_TOKEN_BITBUCKET_SERVER_READ=xx
export GITPOD_TEST_TOKEN_BITBUCKET_SERVER_WRITE=xx
export GITPOD_TEST_TOKEN_BITBUCKET_SERVER_ADMIN=xx
yarn test:unit
```

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - hw-bbs-push-perm</li>
	<li><b>🔗 URL</b> - <a href="https://hw-bbs-push-perm.preview.gitpod-dev.com/workspaces" target="_blank">hw-bbs-push-perm.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - hw-bbs-push-perm-gha.16093</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-hw-bbs-push-perm%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/bitbucket-server/bitbucket-server-token-validator.ts [hunk 1]
```diff
@@ -34,25 +34,18 @@ export class BitbucketServerTokenValidator implements IGitTokenValidator {
...
-            }
+        if (!found) {
+            return {
+                found,
+                isPrivateRepo,
+                writeAccessToRepo,
+            };
         }
 
+        // We can't check REPO_WRITE permission on BitBucketServer since there's no open api for it if we don't change data
```
Comment (AlexTugarev) [inline components/server/src/bitbucket-server/bitbucket-server-token-validator.ts:45]: Yes we can!

You can use the [Search for repositories](https://developer.atlassian.com/server/bitbucket/rest/v811/api-group-repository/#api-api-latest-repos-get) endpoint like this  /rest/api/1.0/repos?permission=REPO_WRITE&name=foobar-read-only to test for write permissions, if the project+repository name are known.

Please change the comment to make clear, that we're doing a short cut here which is good enough as long as we request all scopes at once. If we would to change the scope elevation for BBS, it's better to fix the `getPermission` implementation, but we can proceed with removing for now. 
Screenshot 2023-08-29 at 08 02 58

Comment (mustard-mh) [reply components/server/src/bitbucket-server/bitbucket-server-token-validator.ts]: Changed to use repo search to verify REPO_WRITE permission

**Comment by AlexTugarev** [comment]
@mustard-mh, code looks good to me. 
WDYT should we rely just on manual testing, or write some test?

**Comment by mustard-mh** [comment]
> @mustard-mh, code looks good to me. WDYT should we rely just on manual testing, or write some test?

@AlexTugarev  Yes, I agree that we need unit tests, but the unit tests for BBS in our code base is not that good now, we will need a constant environment:

- Addition admin/write/read permission account for tests only
  - Instead of someone's account
  - Add in 1Password
- Constant repos that don't allow others to edit and change (like push that change ref and new branch etc)
- Better how to run unit tests or just make those environment a part of our gitpod-io/gitpod Project's environment

**Comment by mustard-mh** [comment]
@AlexTugarev I have update permission check test cases, you can try run unit tests. Access Token s are inside 1Password

**Comment by mustard-mh** [comment]
@AlexTugarev could you take a look again? 🙏

**Review by AlexTugarev — APPROVED** [review]

**Pull request merged**